src/sa.h

```diff
@@ -0,0 +1,112 @@
+#include <stdio.h>
+#include <string.h>
+#include <malloc.h>
+#include "parsers/simple-fasta-parser.h"
+#include "parsers/simple-fastq-parser.h"
+#include <stdint-gcc.h>
+#include "helper.h"
+
+int *constructSARadix(struct Fasta fasta) {
+    char* x = fasta.fasta_sequence;
+    int n = fasta.fasta_len;
+    int* sa = malloc(n * sizeof *sa);
+    int* saCopy = malloc(n * sizeof *sa);
+    int** currentSa = &sa;
+    int** otherSa = &saCopy;
+    int** temp;
+
+    for (int i = 0; i<n; i++) {
+        sa[i] = i;
+    }
+    int* bucketsIndecies = fasta.alphabet.sightings;
+    int accumSum = 0;
+    for(int i=0; i<fasta.alphabet.size; i++) {
+        int sighting = bucketsIndecies[i];
+        bucketsIndecies[i] = accumSum;
+        accumSum += sighting;
+    }
+    int* buckets = malloc(fasta.alphabet.size*sizeof *buckets);
+
+    int charIndex;
+    char c;
+    int elemInBucket;
+    int saIndex;
+    for(int i=n-1; i>=0; i--) {
+        for(int j=0; j<fasta.alphabet.size; j++) buckets[j] = 0;
+        for(int j=0; j<n; j++) {
+            charIndex = ((*currentSa)[j] + i) % n;
+            c = x[charIndex];
+            elemInBucket = buckets[c]++;
+            saIndex = bucketsIndecies[c] + elemInBucket;
+            (*otherSa)[saIndex] = (*currentSa)[j];
+        }
+        temp = currentSa;
+        currentSa = otherSa;
+        otherSa = temp;
+    }
+    return *currentSa;
+}
+
+int **constructMultipleSARadix(struct FastaContainer *fastaContainer) {
+    int ** SAs = malloc(fastaContainer->numberOfFastas*sizeof *SAs);
+
+    for(int i=0; i<fastaContainer->numberOfFastas; i++) {
+        int *sa = constructSARadix(*(fastaContainer->fastas)[i]);
+        SAs[i] = sa;
+    }
+    return SAs;
+}
+
+
+struct Interval {
+    int start;
+    int mid;
+    int end;
+};
+
+struct Interval binarySearch(const char* x, const int* sa, char patchar, int parIndex, struct Interval interval, int mode) {
+    char xchar;
+    while (interval.start != interval.end) {
+        interval.mid = (interval.start+interval.end)/2;
+        xchar = x[sa[interval.mid]+parIndex];
```

Comment:
Are you sure you can't get an overflow here? Is it impossible for `sa[interval.mid] + parIndex >= strlen(x)`?

Reply:
Ok, yes then it works